x-pack/filebeat/input/s3/collector.go

```diff
@@ -0,0 +1,635 @@
+				if err != nil {
+					c.logger.Error(errors.Wrap(err, "SQS ChangeMessageVisibilityRequest failed"))
+				}
+				c.logger.Infof("Message visibility timeout updated to %v", visibilityTimeout)
+			} else {
+				// When ACK done, message will be deleted. Or when message is
+				// not s3 ObjectCreated event related(handleSQSMessage function
+				// failed), it will be removed as well.
+				c.logger.Debug("Deleting message from SQS: ", *message.MessageId)
+				// only delete sqs message when errC is closed with no error
+				err := c.deleteMessage(queueURL, *message.ReceiptHandle, svcSQS)
+				if err != nil {
+					c.logger.Error(errors.Wrap(err, "deleteMessages failed"))
+				}
+			}
+			return
+		case <-time.After(time.Duration(visibilityTimeout/2) * time.Second):
+			c.logger.Warn("Half of the set visibilityTimeout passed, visibility timeout needs to be updated")
+			// If half of the set visibilityTimeout passed and this is
+			// still ongoing, then change visibility timeout.
+			err := c.changeVisibilityTimeout(queueURL, visibilityTimeout, svcSQS, message.ReceiptHandle)
+			if err != nil {
+				c.logger.Error(errors.Wrap(err, "SQS ChangeMessageVisibilityRequest failed"))
+			}
+			c.logger.Infof("Message visibility timeout updated to %v seconds", visibilityTimeout)
+		}
+	}
+}
+
+func (c *s3Collector) receiveMessage(svcSQS sqsiface.ClientAPI, visibilityTimeout int64) (*sqs.ReceiveMessageResponse, error) {
+	// receive messages from sqs
+	req := svcSQS.ReceiveMessageRequest(
+		&sqs.ReceiveMessageInput{
+			QueueUrl:              &c.config.QueueURL,
+			MessageAttributeNames: []string{"All"},
+			MaxNumberOfMessages:   &maxNumberOfMessage,
+			VisibilityTimeout:     &visibilityTimeout,
+			WaitTimeSeconds:       &waitTimeSecond,
+		})
+
+	// The Context will interrupt the request if the timeout expires.
+	sendCtx, cancelFn := context.WithTimeout(c.cancellation, c.config.APITimeout)
+	defer cancelFn()
+
+	return req.Send(sendCtx)
+}
+
+func (c *s3Collector) changeVisibilityTimeout(queueURL string, visibilityTimeout int64, svcSQS sqsiface.ClientAPI, receiptHandle *string) error {
+	req := svcSQS.ChangeMessageVisibilityRequest(&sqs.ChangeMessageVisibilityInput{
+		QueueUrl:          &queueURL,
+		VisibilityTimeout: &visibilityTimeout,
+		ReceiptHandle:     receiptHandle,
+	})
+
+	// The Context will interrupt the request if the timeout expires.
+	sendCtx, cancelFn := context.WithTimeout(c.cancellation, c.config.APITimeout)
+	defer cancelFn()
+
+	_, err := req.Send(sendCtx)
+	return err
+}
+
+func getRegionFromQueueURL(queueURL string) (string, error) {
+	// get region from queueURL
+	// Example: https://sqs.us-east-1.amazonaws.com/627959692251/test-s3-logs
+	queueURLSplit := strings.Split(queueURL, ".")
+	if queueURLSplit[0] == "https://sqs" && queueURLSplit[2] == "amazonaws" {
+		return queueURLSplit[1], nil
+	}
+	return "", errors.New("queueURL is not in format: https://sqs.{REGION_ENDPOINT}.amazonaws.com/{ACCOUNT_NUMBER}/{QUEUE_NAME}")
```

Comment:
AFAIK for creating new errors our convention is to use `fmt.Errorf`.

Reply:
Done! Thank you! I also replaced all errors.Wrap with fmt.Errorf